distributed/protocol/serialize.py

```diff
@@ -445,34 +445,46 @@ def extract_serialize(x):
     >>> extract_serialize(msg)
     ({'op': 'update'}, {('data',): <Serialize: 123>}, set())
     """
-    x2 = type(x)()
-    ser = {}
-    bytestrings = set()
-    _extract_serialize(x, x2, ser, bytestrings)
-    return x2, ser, bytestrings
-
-
-def _extract_serialize(x, x2, ser, bytestrings, path=()):
     typ_x = type(x)
     if typ_x is dict:
         x_items = x.items()
+        x2 = {}
     elif typ_x is list:
         x_items = enumerate(x)
-        x2.extend(repeat(None, len(x)))
-
-    for k, v in x_items:
-        path_k = path + (k,)
-        typ_v = type(v)
-        if typ_v is dict or typ_v is list:
-            x2[k] = v2 = typ_v()
-            _extract_serialize(v, v2, ser, bytestrings, path_k)
-        elif typ_v is Serialize or typ_v is Serialized:
-            ser[path_k] = v
-        elif (typ_v is bytes or typ_v is bytearray) and len(v) > 2 ** 16:
-            ser[path_k] = to_serialize(v)
-            bytestrings.add(path_k)
+        x2 = len(x) * [None]
+
+    ser = {}
+    bytestrings = set()
+    _extract_serialize(x_items, x2, ser, bytestrings)
+    return x2, ser, bytestrings
+
+
+def _extract_serialize(x_items, x2, ser, bytestrings, path=()):
+    q = deque()
```

Comment:
If we're using this as only a stack then maybe this should be a list rather than a deque

Reply:
Yeah I tried both. This performed better based on the benchmarking I did. That said, it seems that recursion itself is already handled quite efficiently.